here-naksha-app-service/src/test/java/com/here/naksha/app/service/ValDryRunTest.java

```diff
@@ -0,0 +1,153 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+import com.here.naksha.app.common.ApiTest;
+import com.here.naksha.app.common.NakshaTestWebClient;
+import com.here.naksha.app.service.models.FeatureCollectionRequest;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzFeature;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzFeatureCollection;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzReference;
+import com.here.naksha.lib.core.models.naksha.Space;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.UUID;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONException;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.skyscreamer.jsonassert.JSONAssert;
+import org.skyscreamer.jsonassert.JSONCompareMode;
+import org.skyscreamer.jsonassert.comparator.ArraySizeComparator;
+
+public class ValDryRunTest extends ApiTest {
+
+  private static final NakshaTestWebClient nakshaClient = new NakshaTestWebClient();
+
+  private static final String SPACE_ID = "local-space-4-val-dry-run";
+
+  @BeforeAll
+  static void setup() throws URISyntaxException, IOException, InterruptedException {
+    createHandler(nakshaClient, "ValDryRun/setup");
+    createSpace(nakshaClient, "ValDryRun/setup");
+  }
+
+  private void additionalCustomAssertions_tc3000(final @NotNull String reqBody, final @NotNull String resBody)
+      throws JSONException {
+    final FeatureCollectionRequest collectionRequest = parseJson(reqBody, FeatureCollectionRequest.class);
+    final XyzFeatureCollection collectionResponse = parseJson(resBody, XyzFeatureCollection.class);
+    final List<String> updatedIds = collectionResponse.getUpdated();
+    final List<XyzFeature> features = collectionResponse.getFeatures();
+    final List<XyzFeature> violations = collectionResponse.getViolations();
+    JSONAssert.assertEquals(
+        "{updated:[" + collectionRequest.getFeatures().size() + "]}",
+        resBody,
+        new ArraySizeComparator(JSONCompareMode.LENIENT));
+    assertEquals(
+        updatedIds.size(), features.size(), "Mismatch between updated and features list size in the response");
+    final String newFeatureId = features.get(2).getId();
+    assertNotNull(newFeatureId, "Feature Id must not be null");
+    for (int i = 3; i <= 5; i++) {
+      final XyzFeature violation = violations.get(i);
+      final List<XyzReference> references = violation.getProperties().getReferences();
+      assertNotNull(references, "References missing for violation at idx " + i);
+      for (final XyzReference reference : references) {
+        assertNotNull(reference.getId(), "Id missing in references for violation at idx " + i);
+        assertEquals(newFeatureId, reference.getId(), "Violation referenced featured id doesn't match");
+      }
+    }
+  }
```

Comment:
nth: I wonder if we can extract separate assertions from this big method, making it a bit more readable (I have similar feelings towards other "additionalCustomAssertions" methods as well tbh).
There are couple of things going on here:
1) checking updated size (json-based)
2) comparing features with updated (getter-based)
3) checking feature id
4) explicit violations checks (for only some of features?)

Reply:
Alright. It is on the lines of existing CreateFeatureTest.additionalCustomAssertions() ... will see what we can do about it. But if it takes longer, I will take as separate PR.

Reply:
it's nth - we don't have to address it now :)